File: app/frontend/channels/room_channel.js
import consumer from './consumer'

document.addEventListener('turbolinks:load', () => {
  let messages

  messages = document.getElementById('messages')

  if (messages) {
    createRoomChannel(messages.dataset.roomId)
  }

  const message = document.getElementById('message_body')

  if (message) {
    messages.scrollTop = messages.scrollHeight

    message.addEventListener('keydown', (event) => {
      if (event.key === 'Enter' && message.value.length > 0) {
        room.speak(message.value)
        message.value = ''
        event.preventDefault()
      }
    })
  }
})

let room

const createRoomChannel = (roomId) => {
  room = consumer.subscriptions.create({
    channel: 'RoomChannel',
    roomId: roomId
  }, {

    received(data) {
      messages.insertAdjacentHTML('beforeend', (data['message']))
      messages.scrollTop = messages.scrollHeight
    },

    speak(message) {
      return this.perform('speak', {
        message: message
      })
    }
  })
}

File: app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    @room = Room.find(params[:roomId])

    stream_from "room_channel_#{@room.id}"

    speak('message' => '* * * joined the room * * *')
  end

  def unsubscribed
    speak('message' => '* * * left the room * * *')
  end

  def speak(data)
    MessageService.new(
      body: data['message'],
      room: @room,
      user: current_user
    ).perform
  end
end
